From e51317eb6deea8fb98999fb8549c36b2d42a16b5 Mon Sep 17 00:00:00 2001
From: Milian Wolff <milian.wolff@kdab.com>
Date: Thu, 1 Jul 2021 11:01:48 +0200
Subject: [PATCH 021/163] Fix memory leak of QOffscreenScreen in
 QOffscreenIntegration

Call handleScreenRemoved when the QOffscreenIntegration is getting
destroyed. Fixes memory leaks such as this one:

```
==207317== 368 (40 direct, 328 indirect) bytes in 1 blocks are definitely lost in loss record 157 of 264
==207317==    at 0x483EF3F: operator new(unsigned long) (vg_replace_malloc.c:417)
==207317==    by 0x987C434: QOffscreenIntegration::QOffscreenIntegration() (qoffscreenintegration.cpp:125)
==207317==    by 0x987CE81: QOffscreenX11Integration::QOffscreenX11Integration() (qoffscreenintegration_x11.h:69)
==207317==    by 0x987C8CC: QOffscreenIntegration::createOffscreenIntegration() (qoffscreenintegration.cpp:239)
==207317==    by 0x987BAA8: QOffscreenIntegrationPlugin::create(QString const&, QStringList const&) (main.cpp:58)
==207317==    by 0x5532FEC: QPlatformIntegrationPlugin::create(QString const&, QStringList const&, int&, char**) (qplatformintegrationplugin.cpp:64)
==207317==    by 0x5532AE0: QPlatformIntegration* qLoadPlugin<QPlatformIntegration, QPlatformIntegrationPlugin, QStringList const&, int&, char**&>(QFactoryLoader const*, QString const&, QStringList const&, int&, char**&) (qfactoryloader_p.h:108)
==207317==    by 0x5532301: QPlatformIntegrationFactory::create(QString const&, QStringList const&, int&, char**, QString const&) (qplatformintegrationfactory.cpp:71)
==207317==    by 0x553F502: init_platform(QString const&, QString const&, QString const&, int&, char**) (qguiapplication.cpp:1223)
==207317==    by 0x554280B: QGuiApplicationPrivate::createPlatformIntegration() (qguiapplication.cpp:1481)
==207317==    by 0x554294F: QGuiApplicationPrivate::createEventDispatcher() (qguiapplication.cpp:1498)
==207317==    by 0x4C53D6F: QApplicationPrivate::createEventDispatcher() (qapplication.cpp:167)
```

Upstream has diverged, fix for 6.2 is available here:
https://codereview.qt-project.org/c/qt/qtbase/+/357736

Change-Id: Ife36cb079340bb72f5943e87d621af2f8111c238
---
 src/plugins/platforms/offscreen/qoffscreenintegration.cpp | 4 +++-
 src/plugins/platforms/offscreen/qoffscreenintegration.h   | 1 +
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/plugins/platforms/offscreen/qoffscreenintegration.cpp b/src/plugins/platforms/offscreen/qoffscreenintegration.cpp
index 141fb68a23..d4294d425a 100644
--- a/src/plugins/platforms/offscreen/qoffscreenintegration.cpp
+++ b/src/plugins/platforms/offscreen/qoffscreenintegration.cpp
@@ -122,11 +122,13 @@ QOffscreenIntegration::QOffscreenIntegration()
 #endif
     m_services.reset(new QPlatformServices);
 
-    QWindowSystemInterface::handleScreenAdded(new QOffscreenScreen);
+    m_screen = new QOffscreenScreen;
+    QWindowSystemInterface::handleScreenAdded(m_screen);
 }
 
 QOffscreenIntegration::~QOffscreenIntegration()
 {
+    QWindowSystemInterface::handleScreenRemoved(m_screen);
 }
 
 void QOffscreenIntegration::initialize()
diff --git a/src/plugins/platforms/offscreen/qoffscreenintegration.h b/src/plugins/platforms/offscreen/qoffscreenintegration.h
index 0ea90f6c2f..fe00fde07c 100644
--- a/src/plugins/platforms/offscreen/qoffscreenintegration.h
+++ b/src/plugins/platforms/offscreen/qoffscreenintegration.h
@@ -84,6 +84,7 @@ protected:
 #endif
     QScopedPointer<QPlatformInputContext> m_inputContext;
     QScopedPointer<QPlatformServices> m_services;
+    QPlatformScreen *m_screen;
     mutable QScopedPointer<QPlatformNativeInterface> m_nativeInterface;
 };
 
-- 
2.37.2

